Add Compose hot reload (#383)
JetBrains/kotlin-multiplatform-dev-docs: topics/tools/compose-hot-reload.md

[//]: # (title: Compose hot reload)

<primary-label ref="alpha"/>

Compose hot reload is a feature for your IDE that helps you visualize and experiment with UI changes while
working on a Compose Multiplatform project.

At the moment, Compose hot reload is only available when you include a desktop target in your multiplatform project. 
We're exploring adding support for other targets in the future. In the meantime, using the desktop app as your sandbox

> lets you quickly experiment with UI changes without interrupting your flow.

Reviewer: ```suggestion
lets you quickly experiment with UI changes in common code without interrupting your flow.
```

Maybe it's unnecessary, but I'd like to stress that the desktop target helps you work with common code, it's not like with Compose previews that only worked for Android.

## Use Compose hot reload

1. In the `desktopMain` directory, open the `main.kt` file and update the `main()` function:
   ```kotlin
   fun main() = application {
       Window(
           onCloseRequest = ::exitApplication,
           alwaysOnTop = true,
           title = "composedemo",
       ) {
           App()
       }
   }
   ```
   By setting the `alwaysOnTop` variable to `true`, the generated desktop app stays on top of all your windows, making it easier
   to edit your code and see changes live.

2. In the `commonMain` directory, open the `App.kt` file and update the `Button` composable:
   ```kotlin
   Button(onClick = { showContent = !showContent }) {
       Column {
           Text(Greeting().greet())
       }
   }
   ```
   Now, the text for the button is controlled by the `greet()` function.

3. In the `commonMain` directory, open the `Greeting.kt` file and update the `greet()` function:
   ```kotlin
    fun greet(): String {
        return "Hello!"
    }
   ```

4.  In the `desktopMain` directory, open the `main.kt` file and click the **Run** icon in the gutter. 

>     Select **Run 'main [desktop]' with Compose Hot Reload (Alpha)**. The desktop app is generated.

Reviewer: I'm not sure what it means that "the desktop app is generated", and whether it's useful information. Should we just leave it at "Select Run..."?

    ![Run Compose hot reload from gutter](compose-hot-reload-gutter-run.png){width=350}

    ![First Compose hot reload on desktop app](compose-hot-reload-hello.png){width=500}

5. Update the string returned from the `greet()` function and see the desktop app update automatically.

   ![Compose hot reload](compose-hot-reload.gif){width=500}

> Congratulations! You've run your first Compose hot reload. Now you can experiment with changing text, images, formatting, 

Reviewer: ```suggestion
Congratulations! You've seen Compose hot reload in action. Now you can experiment with changing text, images, formatting, 
```

> UI structure, and more, without having to restart the desktop run configuration every time.

Reviewer: ```suggestion
UI structure, and more, without having to restart the desktop run configuration after every change.
```
